File: test/transitionPolicy.test.js
import assert from 'node:assert/strict';
import test from 'node:test';

import {
  assessTransitionTier,
  audibleSecondsBetween,
  isVocalClash,
  rankMixInCandidates,
  resolveMixOutAnchor,
  vocalActivityBetween
} from '../src/audio/crossfade/transitionPolicy.js';

function energyCurve(duration, silentFrom = Infinity) {
  const curve = [];
  for (let time = 0; time < duration; time += 1) {
    curve.push({ time, energy: time >= silentFrom ? 0 : 0.8 });
  }
  return curve;
}

function maskedAnalysis(values, secondsPerSample = 1) {
  return {
    energyCurve: values.map((_, index) => ({ time: index * secondsPerSample, energy: 0.8 })),
    vocalActivityMask: values
  };
}

test('trusted grids within the stretch window rate a beat-matched transition', () => {
  const policy = assessTransitionTier({
    analysis: { bpm: 126, beatConfidence: 0.9 },
    nextAnalysis: { bpm: 124, beatConfidence: 0.7 }
  });
  assert.equal(policy.tier, 'beatmatched');
  assert.deepEqual(policy.reasons, []);
  assert.equal(policy.beatConfidence, 0.7);
});

test('octave-distant tempi are counted on the shared grid before judging', () => {
  const policy = assessTransitionTier({
    analysis: { bpm: 126, beatConfidence: 0.9 },
    nextAnalysis: { bpm: 63, beatConfidence: 0.9 }
  });
  assert.equal(policy.tier, 'beatmatched');
});

test('tempo distance demotes to the DJ-assisted tier', () => {
  const policy = assessTransitionTier({
    analysis: { bpm: 126, beatConfidence: 0.9 },
    nextAnalysis: { bpm: 100, beatConfidence: 0.9 }
  });
  assert.equal(policy.tier, 'dj_assisted');
  assert.equal(policy.reasons[0], 'tempo-distance');
});

test('one weak beat grid demotes to the DJ-assisted tier', () => {
  const policy = assessTransitionTier({
    analysis: { bpm: 126, beatConfidence: 0.9 },
    nextAnalysis: { bpm: 126, beatConfidence: 0.3 }
  });
  assert.equal(policy.tier, 'dj_assisted');
  assert.deepEqual(policy.reasons, ['beat-confidence']);
});

test('missing tempo or two untrusted grids bottom out at a plain crossfade', () => {
  assert.equal(
    assessTransitionTier({
      analysis: { bpm: 0, beatConfidence: 0.9 },
      nextAnalysis: { bpm: 126, beatConfidence: 0.9 }
    }).tier,
    'plain_crossfade'
  );
  const catalogOnly = assessTransitionTier({
    analysis: { bpm: 120, beatConfidence: 0 },
    nextAnalysis: { bpm: 120, beatConfidence: 0.1 }
  });
  assert.equal(catalogOnly.tier, 'plain_crossfade');
  assert.deepEqual(catalogOnly.reasons, ['beat-confidence']);
});

test('audible seconds count music and ignore silence', () => {
  const analysis = { energyCurve: energyCurve(100, 60) };
  assert.ok(Math.abs(audibleSecondsBetween(analysis, 0, 59) - 60) < 1.5);
  assert.equal(audibleSecondsBetween(analysis, 60, 99), 0);
  assert.equal(audibleSecondsBetween({}, 0, 10), null);
});

test('an outro marker never anchors a transition ahead of real music', () => {
  // The native analyzer marks an outro up to 48s before content end.
  const anchor = resolveMixOutAnchor({
    contentEndTime: 240,
    outroStartTime: 192,
    energyCurve: energyCurve(240)
  });
  assert.equal(anchor.type, 'content_end');
  assert.equal(anchor.time, 240);
});

test('a genuine silence cliff still anchors the transition', () => {
  const anchor = resolveMixOutAnchor({
    contentEndTime: 264,
    mixOutTime: 188,
    energyCurve: energyCurve(264, 188)
  });
  assert.equal(anchor.type, 'interior_mix_out');
  assert.equal(anchor.time, 188);
  assert.equal(anchor.discardedMusicSeconds, 0);
});

test('a silence cliff followed by more music is not a mix-out', () => {
  // A four-second gap, then the final chorus. Skipping here would cut the
  // song short even though the analyzer scored the boundary highly.
  const curve = energyCurve(264).map((point) => (
    point.time >= 188 && point.time < 192 ? { ...point, energy: 0 } : point
  ));
  const anchor = resolveMixOutAnchor({
    contentEndTime: 264,
    mixOutTime: 188,
    energyCurve: curve
  });
  assert.equal(anchor.type, 'content_end');
});

test('a short unplayed tail stays within the budget', () => {
  const anchor = resolveMixOutAnchor({
    contentEndTime: 200,
    mixOutTime: 192,
    energyCurve: energyCurve(200)
  });
  assert.equal(anchor.type, 'interior_mix_out');
  assert.ok(anchor.discardedMusicSeconds <= 12);
});

test('without an energy curve the budget errs toward playing the track', () => {
  assert.equal(resolveMixOutAnchor({ contentEndTime: 240, mixOutTime: 180 }).type, 'content_end');
  assert.equal(resolveMixOutAnchor({ contentEndTime: 240, mixOutTime: 232 }).type, 'interior_mix_out');
});

test('vocal activity averages only the samples inside the window', () => {
  const analysis = maskedAnalysis([0.9, 0.9, 0.9, 0.1, 0.1, 0.1]);
  assert.equal(vocalActivityBetween(analysis, 0, 2), 0.9);
  assert.ok(Math.abs(vocalActivityBetween(analysis, 3, 5) - 0.1) < 1e-9);
  assert.equal(vocalActivityBetween(analysis, 10, 20), null);
  assert.equal(vocalActivityBetween({}, 0, 10), null);
  assert.equal(vocalActivityBetween(analysis, 5, 5), null);
});

test('a vocal clash needs measurable singing on both sides', () => {
  assert.equal(isVocalClash(0.9, 0.8), true);
  assert.equal(isVocalClash(0.9, 0.2), false);
  // The JS fallback analyzer's flat 0.5 mask must never trip vocal logic.
  assert.equal(isVocalClash(0.5, 0.5), false);
  assert.equal(isVocalClash(null, 0.9), false);
});

test('an analyzed drop outranks a higher-scored phrase boundary', () => {
  const ranked = rankMixInCandidates({
    bpm: 126,
    mixInCandidates: [
      { time: 45, score: 0.4, type: 'phrase' },
      { time: 30.476, score: 0.2, type: 'main_drop' }
    ]
  });
  assert.equal(ranked[0].time, 30.476);
});

test('a vocal run-up demotes an otherwise equal entry point', () => {
  const beat = 60 / 126;
  const mask = [];
  const curve = [];
  for (let time = 0; time < 80; time += 1) {
    curve.push({ time, energy: 0.8 });
    // Singing leads into the first candidate; the second's run-up is clear.
    mask.push(time < 32 ? 0.9 : 0.1);
  }
  const ranked = rankMixInCandidates({
    bpm: 126,
    energyCurve: curve,
    vocalActivityMask: mask,
    mixInCandidates: [
      { time: 32 * beat * 2, score: 0.9, type: 'intro_drop' },
      { time: 64 * beat * 2, score: 0.9, type: 'intro_drop' }
    ]
  });
  assert.equal(ranked[0].time, 64 * beat * 2);
});

test('a cold open with no run-up is penalized', () => {
  const ranked = rankMixInCandidates({
    bpm: 120,
    audibleStartTime: 0,
    mixInCandidates: [
      { time: 0.5, score: 0.8, type: 'pickup' },
      { time: 16, score: 0.8, type: 'pickup' }
    ]
  });
  assert.equal(ranked[0].time, 16);
});

File: mobile/android/app/src/test/java/dev/sfg/orchard/mobile/playback/smart/TransitionPolicyTest.kt
/*
 * Copyright (C) 2026 SFG545
 *
 * This file is part of Orchard.
 *
 * Orchard is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option) any
 * later version.
 *
 * Orchard is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with Orchard. If not, see <https://www.gnu.org/licenses/>.
 */

package dev.sfg.orchard.mobile.playback.smart

import kotlin.math.abs
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/** Ported from Orchard desktop's `test/transitionPolicy.test.js`, expectations included. */
class TransitionPolicyTest {

    private fun energyCurve(duration: Int, silentFrom: Int = Int.MAX_VALUE): List<EnergySample> =
        (0 until duration).map { time ->
            EnergySample(time.toDouble(), if (time >= silentFrom) 0.0 else 0.8)
        }

    private fun maskedAnalysis(values: List<Double>, secondsPerSample: Double = 1.0) = TrackAnalysis(
        energyCurve = values.indices.map { EnergySample(it * secondsPerSample, 0.8) },
        vocalActivityMask = values,
    )

    @Test
    fun `trusted grids within the stretch window rate a beat-matched transition`() {
        val policy = assessTransitionTier(
            TrackAnalysis(bpm = 126.0, beatConfidence = 0.9),
            TrackAnalysis(bpm = 124.0, beatConfidence = 0.7),
        )
        assertEquals(TransitionTier.BEATMATCHED, policy.tier)
        assertEquals(emptyList<String>(), policy.reasons)
        assertEquals(0.7, policy.beatConfidence, 1e-9)
    }

    @Test
    fun `octave-distant tempi are counted on the shared grid before judging`() {
        val policy = assessTransitionTier(
            TrackAnalysis(bpm = 126.0, beatConfidence = 0.9),
            TrackAnalysis(bpm = 63.0, beatConfidence = 0.9),
        )
        assertEquals(TransitionTier.BEATMATCHED, policy.tier)
    }

    @Test
    fun `tempo distance demotes to the DJ-assisted tier`() {
        val policy = assessTransitionTier(
            TrackAnalysis(bpm = 126.0, beatConfidence = 0.9),
            TrackAnalysis(bpm = 100.0, beatConfidence = 0.9),
        )
        assertEquals(TransitionTier.DJ_ASSISTED, policy.tier)
        assertEquals("tempo-distance", policy.reasons.first())
    }

    @Test
    fun `one weak beat grid demotes to the DJ-assisted tier`() {
        val policy = assessTransitionTier(
            TrackAnalysis(bpm = 126.0, beatConfidence = 0.9),
            TrackAnalysis(bpm = 126.0, beatConfidence = 0.3),
        )
        assertEquals(TransitionTier.DJ_ASSISTED, policy.tier)
        assertEquals(listOf("beat-confidence"), policy.reasons)
    }

    @Test
    fun `missing tempo or two untrusted grids bottom out at a plain crossfade`() {
        assertEquals(
            TransitionTier.PLAIN_CROSSFADE,
            assessTransitionTier(
                TrackAnalysis(bpm = 0.0, beatConfidence = 0.9),
                TrackAnalysis(bpm = 126.0, beatConfidence = 0.9),
            ).tier,
        )
        // A catalog BPM lookup merges in at beatConfidence 0, so metadata alone can never
        // authorize beat-matching however plausible the tempo looks.
        val catalogOnly = assessTransitionTier(
            TrackAnalysis(bpm = 120.0, beatConfidence = 0.0),
            TrackAnalysis(bpm = 120.0, beatConfidence = 0.1),
        )
        assertEquals(TransitionTier.PLAIN_CROSSFADE, catalogOnly.tier)
        assertEquals(listOf("beat-confidence"), catalogOnly.reasons)
    }

    @Test
    fun `audible seconds count music and ignore silence`() {
        val analysis = TrackAnalysis(energyCurve = energyCurve(100, 60))
        assertTrue(abs(audibleSecondsBetween(analysis, 0.0, 59.0)!! - 60) < 1.5)
        assertEquals(0.0, audibleSecondsBetween(analysis, 60.0, 99.0)!!, 1e-9)
        assertNull(audibleSecondsBetween(TrackAnalysis(), 0.0, 10.0))
    }

    @Test
    fun `an outro marker never anchors a transition ahead of real music`() {
        // The native analyzer marks an outro up to 48s before content end.
        val anchor = resolveMixOutAnchor(
            TrackAnalysis(
                contentEndTime = 240.0,
                outroStartTime = 192.0,
                energyCurve = energyCurve(240),
            ),
        )
        assertEquals("content_end", anchor.type)
        assertEquals(240.0, anchor.time, 1e-9)
    }

    @Test
    fun `a genuine silence cliff still anchors the transition`() {
        val anchor = resolveMixOutAnchor(
            TrackAnalysis(
                contentEndTime = 264.0,
                mixOutTime = 188.0,
                energyCurve = energyCurve(264, 188),
            ),
        )
        assertEquals("interior_mix_out", anchor.type)
        assertEquals(188.0, anchor.time, 1e-9)
        assertEquals(0.0, anchor.discardedMusicSeconds, 1e-9)
    }

    @Test
    fun `a silence cliff followed by more music is not a mix-out`() {
        // A four-second gap, then the final chorus. Skipping here would cut the song short even
        // though the analyzer scored the boundary highly.
        val curve = energyCurve(264).map {
            if (it.time >= 188 && it.time < 192) it.copy(energy = 0.0) else it
        }
        val anchor = resolveMixOutAnchor(
            TrackAnalysis(contentEndTime = 264.0, mixOutTime = 188.0, energyCurve = curve),
        )
        assertEquals("content_end", anchor.type)
    }

    @Test
    fun `an analysis with no energy curve charges the whole gap against the budget`() {
        val anchor = resolveMixOutAnchor(
            TrackAnalysis(contentEndTime = 240.0, mixOutTime = 190.0),
        )
        assertEquals("content_end", anchor.type)
    }

    @Test
    fun `vocal activity is averaged over the window and absent masks read as no evidence`() {
        val analysis = maskedAnalysis(listOf(0.0, 0.0, 1.0, 1.0))
        assertEquals(0.5, vocalActivityBetween(analysis, 0.0, 3.0)!!, 1e-9)
        assertEquals(1.0, vocalActivityBetween(analysis, 2.0, 3.0)!!, 1e-9)
        assertNull(vocalActivityBetween(TrackAnalysis(), 0.0, 3.0))
    }

    @Test
    fun `a clash needs evidence on both sides`() {
        assertTrue(isVocalClash(0.9, 0.8))
        assertTrue(!isVocalClash(0.9, null))
        assertTrue(!isVocalClash(null, null))
        assertTrue(!isVocalClash(0.9, 0.2))
    }

    @Test
    fun `drops outrank phrase lines as entry points`() {
        val ranked = rankMixInCandidates(
            TrackAnalysis(
                bpm = 120.0,
                audibleStartTime = 0.0,
                mixInCandidates = listOf(
                    MixCandidate(time = 16.0, score = 0.5, type = "phrase"),
                    MixCandidate(time = 32.0, score = 0.5, type = "main_drop"),
                ),
            ),
        )
        assertEquals("main_drop", ranked.first().type)
        assertEquals(32.0, ranked.first().time, 1e-9)
    }

    @Test
    fun `a cold open is penalized against an entry with an intro to bed under`() {
        val analysis = TrackAnalysis(
            bpm = 120.0,
            audibleStartTime = 0.0,
            mixInCandidates = listOf(
                MixCandidate(time = 0.5, score = 0.6, type = "main_drop"),
                MixCandidate(time = 32.0, score = 0.5, type = "main_drop"),
            ),
        )
        assertEquals(32.0, rankMixInCandidates(analysis).first().time, 1e-9)
    }

    @Test
    fun `tempo octave alignment folds a half-time reading onto the shared grid`() {
        assertEquals(126.0, alignTempoOctave(126.0, 63.0), 1e-9)
        assertEquals(126.0, alignTempoOctave(126.0, 252.0), 1e-9)
        assertEquals(124.0, alignTempoOctave(126.0, 124.0), 1e-9)
    }
}
